A race between FelixStartLevel and AutoDeployer threads

### GlassFish Version (and build number)

Glassiish 7.0.25 (and earlier)

### JDK version

Temurin-17.0.14+7 or earlier

### OS

unix or windows

### Database

_No response_

### Problem Description

I found a random rare startup issue on Glassfish 7.
First the original error from the logs:
```
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.IllegalStateException: RoleMapper registration failed! [in thread "FelixStartLevel"]
	at org.glassfish.exousia.modules.locked.SimplePolicyConfiguration.<clinit>(SimplePolicyConfiguration.java:109)
	at org.glassfish.exousia.modules.locked.SimplePolicyProvider.doImplies(SimplePolicyProvider.java:159)
	at org.glassfish.exousia.modules.locked.SimplePolicyProvider.implies(SimplePolicyProvider.java:150)
	at java.base/java.security.ProtectionDomain.implies(ProtectionDomain.java:325)
	at java.base/java.security.ProtectionDomain.impliesWithAltFilePerm(ProtectionDomain.java:357)
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:463)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
	at org.glassfish.security.services.common.SecurityAccessValidator.checkPerm(SecurityAccessValidator.java:256)
	at org.glassfish.security.services.common.SecurityAccessValidator.validateLookup(SecurityAccessValidator.java:164)
	at org.glassfish.security.services.common.SecurityAccessValidator.validate(SecurityAccessValidator.java:59)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.callValidate(ServiceLocatorImpl.java:230)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.validate(ServiceLocatorImpl.java:274)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getDescriptors(ServiceLocatorImpl.java:346)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getDescriptors(ServiceLocatorImpl.java:390)
	at org.glassfish.hk2.extras.events.internal.DefaultTopicDistributionService.configurationChanged(DefaultTopicDistributionService.java:539)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.callAllConfigurationListeners(ServiceLocatorImpl.java:2124)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.addConfiguration(ServiceLocatorImpl.java:2191)
	at org.jvnet.hk2.internal.DynamicConfigurationImpl.commit(DynamicConfigurationImpl.java:265)
	at org.jvnet.hk2.osgiadapter.HK2Main$HK2ServiceTrackerCustomizer.addingService(HK2Main.java:368)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:944)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:872)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:904)
	at org.apache.felix.framework.EventDispatcher.invokeServiceListenerCallback(EventDispatcher.java:990)
	at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:838)
	at org.apache.felix.framework.EventDispatcher.fireServiceEvent(EventDispatcher.java:545)
	at org.apache.felix.framework.Felix.fireServiceEvent(Felix.java:4863)
	at org.apache.felix.framework.Felix.registerService(Felix.java:3834)
	at org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:328)
	at org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:302)
	at org.apache.felix.gogo.runtime.activator.Activator.start(Activator.java:83)
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:849)
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2429)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2335)
	at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1566)
	at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:297)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

which sadly does not tell what is wrong.
Now after some debugging I found out the real issue:

```
2025-06-12 12:22:40.997 INFO  [FelixStartLevel] [GlassFish] Loading: com.sun.enterprise.security.ee.web.integration.GlassfishRoleMapper in jdk.internal.loader.ClassLoaders$AppClassLoader@6073f712
2025-06-12 12:22:40.997 ERROR [FelixStartLevel] [GlassFish] Initializer failure: java.lang.ClassNotFoundException: com.sun.enterprise.security.ee.web.integration.GlassfishRoleMapper:
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at org.glassfish.exousia.modules.locked.SimplePolicyConfiguration.<clinit>(SimplePolicyConfiguration.java:94)
	at org.glassfish.exousia.modules.locked.SimplePolicyProvider.doImplies(SimplePolicyProvider.java:159)
	at org.glassfish.exousia.modules.locked.SimplePolicyProvider.implies(SimplePolicyProvider.java:150)
	at java.base/java.security.ProtectionDomain.implies(ProtectionDomain.java:325)
	at java.base/java.security.ProtectionDomain.impliesWithAltFilePerm(ProtectionDomain.java:357)
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:463)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
	at org.glassfish.security.services.common.SecurityAccessValidator.checkPerm(SecurityAccessValidator.java:256)
	at org.glassfish.security.services.common.SecurityAccessValidator.validateLookup(SecurityAccessValidator.java:164)
	at org.glassfish.security.services.common.SecurityAccessValidator.validate(SecurityAccessValidator.java:59)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.callValidate(ServiceLocatorImpl.java:230)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.validate(ServiceLocatorImpl.java:274)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getDescriptors(ServiceLocatorImpl.java:346)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getDescriptors(ServiceLocatorImpl.java:390)
	at org.glassfish.hk2.extras.events.internal.DefaultTopicDistributionService.configurationChanged(DefaultTopicDistributionService.java:539)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.callAllConfigurationListeners(ServiceLocatorImpl.java:2124)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.addConfiguration(ServiceLocatorImpl.java:2191)
	at org.jvnet.hk2.internal.DynamicConfigurationImpl.commit(DynamicConfigurationImpl.java:265)
	at org.jvnet.hk2.osgiadapter.HK2Main$HK2ServiceTrackerCustomizer.addingService(HK2Main.java:368)
	at OSGi//org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:944)
	at OSGi//org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:872)
	at OSGi//org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at OSGi//org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at OSGi//org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:904)
	at OSGi//org.apache.felix.framework.EventDispatcher.invokeServiceListenerCallback(EventDispatcher.java:990)
	at OSGi//org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:838)
	at OSGi//org.apache.felix.framework.EventDispatcher.fireServiceEvent(EventDispatcher.java:545)
	at OSGi//org.apache.felix.framework.Felix.fireServiceEvent(Felix.java:4863)
	at OSGi//org.apache.felix.framework.Felix.registerService(Felix.java:3834)
	at OSGi//org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:328)
	at OSGi//org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:302)
	at org.apache.felix.gogo.runtime.activator.Activator.newProcessor(Activator.java:76)
	at org.apache.felix.gogo.runtime.activator.Activator.start(Activator.java:85)
	at OSGi//org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:849)
	at OSGi//org.apache.felix.framework.Felix.activateBundle(Felix.java:2429)
	at OSGi//org.apache.felix.framework.Felix.startBundle(Felix.java:2335)
	at OSGi//org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1566)
	at OSGi//org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:297)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

But why does it work on some cases? So I debug it out:

```
2025-06-12 12:05:03.575 INFO  [AutoDeployer] [GlassFish] Loading: com.sun.enterprise.security.ee.web.integration.GlassfishRoleMapper in org.glassfish.javaee.full.deployment.EarClassLoader@55b241a5: java.lang.RuntimeException
	at org.glassfish.exousia.modules.locked.SimplePolicyConfiguration.<clinit>(SimplePolicyConfiguration.java:94)
	at org.glassfish.exousia.modules.locked.SimplePolicyProvider.refresh(SimplePolicyProvider.java:199)
	at com.sun.enterprise.security.PolicyLoader.loadPolicy(PolicyLoader.java:150)
	at com.sun.enterprise.security.SecurityLifecycle.onInitialization(SecurityLifecycle.java:127)
	at com.sun.enterprise.security.SecurityLifecycle.postConstruct(SecurityLifecycle.java:153)
	at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:356)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:410)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:479)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:61)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:49)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
	at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
	at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:100)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2111)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:98)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:68)
	at com.sun.enterprise.security.ee.SecuritySniffer.setup(SecuritySniffer.java:106)
	at com.sun.enterprise.v3.server.ContainerStarter.startContainer(ContainerStarter.java:69)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.setupContainer(ApplicationLifecycle.java:1039)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.setupContainerInfos(ApplicationLifecycle.java:758)
	at org.glassfish.javaee.full.deployment.EarDeployer.prepareBundle(EarDeployer.java:274)
	at org.glassfish.javaee.full.deployment.EarDeployer$1.doBundle(EarDeployer.java:143)
	at org.glassfish.javaee.full.deployment.EarDeployer$1.doBundle(EarDeployer.java:139)
	at org.glassfish.javaee.full.deployment.EarDeployer.doOnBundles(EarDeployer.java:210)
	at org.glassfish.javaee.full.deployment.EarDeployer.doOnAllTypedBundles(EarDeployer.java:218)
	at org.glassfish.javaee.full.deployment.EarDeployer.doOnAllBundles(EarDeployer.java:246)
	at org.glassfish.javaee.full.deployment.EarDeployer.prepare(EarDeployer.java:139)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.prepareModule(ApplicationLifecycle.java:978)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:466)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:256)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:474)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:570)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:566)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:376)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:565)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:596)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:588)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:376)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:587)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1480)
	at com.sun.enterprise.v3.admin.CommandRunnerExecutionContext.execute(CommandRunnerExecutionContext.java:251)
	at org.glassfish.deployment.autodeploy.AutoOperation.run(AutoOperation.java:140)
	at org.glassfish.deployment.autodeploy.AutoDeployer.deploy(AutoDeployer.java:569)
	at org.glassfish.deployment.autodeploy.AutoDeployer.deployAll(AutoDeployer.java:456)
	at org.glassfish.deployment.autodeploy.AutoDeployer.run(AutoDeployer.java:384)
	at org.glassfish.deployment.autodeploy.AutoDeployer.run(AutoDeployer.java:375)
	at org.glassfish.deployment.autodeploy.AutoDeployService$1.run(AutoDeployService.java:208)
	at java.base/java.util.TimerThread.mainLoop(Timer.java:566)
	at java.base/java.util.TimerThread.run(Timer.java:516)
```
	
So in short if GlassfishRoleMapper is loaded fast enough in AutoDeployer thread with the right context classloader then the deploy succeed
but if FelixStartLevel thread is faster to start then AutoDeployer to finish, then it will load GlassfishRoleMapper with the wrong context classloader and the deploy will fail.

### Steps to reproduce

Add some sleeps in case you are running in AutoDeployer thread just before SimplePolicyProvider.refresh is called.

### Impact of Issue

The applications don't deploy on server.

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/PolicyLoader.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/PolicyLoader.java
@@ -125,11 +125,19 @@ public class PolicyLoader {
 
                 boolean usePolicyProxy = Boolean.parseBoolean(System.getProperty(POLICY_PROXY, "true"));
 
-                Policy policy = null;
+                Policy policy;
                 if (usePolicyProxy && System.getSecurityManager() != null) {
                     policy = loadPolicyAsProxy(javaPolicyClassName);
                 } else {
                     policy = loadPolicy(javaPolicyClassName);
+                }
+
+                // TODO: causing ClassCircularity error when SM ON and
+                // deployment use library feature and ApplibClassLoader
+                // it is likely a problem caused by the way classloading is done
+                // in this case.
+                if (System.getSecurityManager() == null) {
+                    policy.refresh();
                 }
 
                 try {
@@ -139,14 +147,6 @@ public class PolicyLoader {
 
                     Method setPolicyMethod = authorizationServiceClass.getMethod("setPolicy", Policy.class);
                     setPolicyMethod.invoke(null, policy);
-                }
-
-                // TODO: causing ClassCircularity error when SM ON and
-                // deployment use library feature and ApplibClassLoader
-                // it is likely a problem caused by the way classloading is done
-                // in this case.
-                if (System.getSecurityManager() == null) {
-                    policy.refresh();
                 }
 
             } catch (Exception e) {

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/PolicyLoader.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/PolicyLoader.java
@@ -132,13 +132,7 @@ public class PolicyLoader {
                     policy = loadPolicy(javaPolicyClassName);
                 }
 
-                // TODO: causing ClassCircularity error when SM ON and
-                // deployment use library feature and ApplibClassLoader
-                // it is likely a problem caused by the way classloading is done
-                // in this case.
-                if (System.getSecurityManager() == null) {
-                    policy.refresh();
-                }
+                policy.refresh();
 
                 try {
                     Policy.setPolicy(policy);

